src/blob/index/core.rs

```diff
@@ -93,16 +94,56 @@ impl<FileIndex: FileIndexTrait> IndexStruct<FileIndex> {
         self.range_filter.clear();
     }
 
-    pub(crate) fn check_filters_key(&self, key: &[u8]) -> FilterResult {
+    pub fn offload_filter(&mut self) {
+        if self.on_disk() {
+            self.bloom_filter.offload_from_memory();
+        }
+    }
+
+    pub(crate) async fn check_filters_key(&self, key: &[u8]) -> FilterResult {
         if !self.range_filter.contains(key) {
             FilterResult::NotContains
-        } else if self.params.bloom_is_on && !self.bloom_filter.contains(key) {
+        } else if self.params.bloom_is_on
+            && matches!(self.check_bloom_key(key).await, Ok(Some(false)))
+        {
             FilterResult::NotContains
         } else {
             FilterResult::NeedAdditionalCheck
         }
     }
 
+    pub(crate) fn check_filters_in_memory(&self, key: &[u8]) -> FilterResult {
+        if !self.range_filter.contains(key) {
+            FilterResult::NotContains
+        } else if self.params.bloom_is_on
+            && self.bloom_filter.contains_in_memory(key) == Some(false)
+        {
+            FilterResult::NotContains
+        } else {
+            FilterResult::NeedAdditionalCheck
+        }
+    }
+
+    pub async fn check_bloom_key(&self, key: &[u8]) -> Result<Option<bool>> {
+        if self.params.bloom_is_on {
+            if let Some(result) = self.bloom_filter.contains_in_memory(key) {
+                Ok(Some(result))
+            } else {
+                Ok(Some(self.bloom_filter.contains_in_file(self, key).await?))
```

Comment:
If understand correctly, you check file on every negative response from filter in memory. What if filter loaded and returned negative?

Reply:
No, `contains_in_memory` returns `None` if the filter is offloaded and the file is checked only in this case